### ISSUE TITLE
Add PAL API intended to expose ARM64/ARM32 Hardware Capabilites

### ISSUE THREAD

**Issue by sdmaclea** [dotnet/coreclr#15421]
JIT will need to know which parts of the ARM64 and/or ARM32 instruction set are available.  Immediate need is for hardware intrinsics, but will be driven by other issues as well.

Linux provides this information in the auxillary vectors.  For `Arm64` use of `getauxval(AT_HWCAP)` with the flags from `#include <asm/hwcap.h>` makes this determination simple.  Note that this list is growing with each kernel as ARM publishes new ARM64 specs. [linux v4.15-rc2 asm/hwcap.h](http://elixir.free-electrons.com/linux/v4.15-rc2/source/arch/arm64/include/uapi/asm/hwcap.h#L43)

Windows provides the [IsProcessorFeaturePresent()](https://msdn.microsoft.com/en-us/library/windows/desktop/ms724482(v=vs.85).aspx) API.  Which is documented for ARM, but not for ARM64.  It looks fairly limited at the moment.

It is unclear how, MacOS will handles this.

Given that 
+ Windows ARM64 API is not public
+ Windows API and Linux API will be updated at different rates
+ MacOS is unclear
+ VM should be OS agnostic, and should not include target specific include files

PAL needs to add an method to expose this info.  

`GetSystemInfo()` is not adequate.  It does not populate CPU info.  The data structure is inadequate to document the non-linear nature of ARM64 instruction set extension.  It is not extensible because it is tied to the Windows API.

`IsProcessorFeaturePresent()` is controlled by Windows API.  Additionally as mentioned above the ARM64 version of this API is not yet public.  It is best to create a new independent API.

Proposal is to add a new PAL API to support this feature.  Function name and returned data structure would be kept orthogonal to all underlying OS's.  It would use a mechanism similar to the JIT EE Interface.  

```C#
PalCapabilityFlags GetPalCpuCapabilities()

class PalCapabilityFlags
{
public:
    enum CapabilityFlag
    {
        PAL_FLAG_HAS_AES                 = 0, // ID_AA64ISAR0_EL1.AES is 1 or better
        PAL_FLAG_HAS_ATOMICS             = 1, // ID_AA64ISAR0_EL1.Atomic is 2 or better
        PAL_FLAG_HAS_CRC32               = 2, // ID_AA64ISAR0_EL1.CRC32 is 1 or better
        PAL_FLAG_HAS_DCPOP               = 3, // ID_AA64ISAR1_EL1.DPB is 1 or better
        PAL_FLAG_HAS_DP                  = 4, // ID_AA64ISAR0_EL1.DP is 1 or better
...
    }

    // Rest of API from JitFlags and/or CORJIT_FLAGS
  }
```

I do not think there is a binary compatibility issue, so I have used the simpler return a `class` approach.

@janvorli @jkotas @tannergooding PTAL  I will start drafting a patch to implement the Arm64/Linux portion of the proposed API.

**Comment by jkotas** [comment]
`PAL_FLAG_HAS_AES                 = 43, // ID_AA64ISAR0_EL1.AES is 1 or better`

Why does this start with 43?

**Comment by sdmaclea** [comment]
> Why does this start with 43?

Because I was editing on Github and I forgot to fix it. I'll correct it above.

**Comment by jkotas** [comment]
Would it make sense to just create the exact set of bits that the JIT/EE interface expects so that we can skip these boring layers that convert one set of bits to a different set of bits?

> I do not think there is a binary compatibility issue, so I have uesd the simpler return a class approach.

How is this simpler?

**Comment by tannergooding** [comment]
I think the difference is that, unlike X86 where CPUID is available to applications, the CPU registers [on ARM] that expose these capabilities are not available to anything but the system (AFAIK).

Each OS already exposes these bits in varying manners, so the exposing them consistently through PAL makes sense.

**Comment by sdmaclea** [comment]
> Would it make sense to just create the exact set of bits

That would be my default approach because I hate these bucket brigades.
We would probably still need to add boring assert code to make sure the bits are in the right place.

Ultimately, I think that would be up to the PAL architects as I do not fully understand everywhere PAL is/will be used.

> How is this simpler?

It may not be.  The JitFlag design needed to runtime check the structs had binary compatibility with CORJIT_FLAGS.  So client created the struct.  Server checked the structure length and asserted if it was different.  

In this case since VM and PAL can use the same header it should be simpler than the VM/JIT case.

I can keep the function call API exactly the same if anyone cares.

**Comment by sdmaclea** [comment]
@jkotas said
> Would it make sense to just create the exact set of bits that the JIT/EE interface expects so that we can skip these boring layers that convert one set of bits to a different set of bits?

@BruceForstall Could we move some part of the corjit.h header into a pal header?

Maybe corjit.h Could just have `CORJIT_FLAG_CPU_FLAG_BEGIN` and `CORJIT_FLAG_CPU_FLAG_END`.  Pal header would need to be kept in sync with jitee.h

**Comment by jkotas** [comment]
> Could we move some part of the corjit.h header into a pal header?

corjit.h is used in number of places. It needs to stay standalone. We should duplicate the bits, or figure out how to #include corjit.h from the PAL impl.

**Comment by sdmaclea** [comment]
> figure out how to #include corjit.h from the PAL impl.

If that is acceptable

I'll try to define the new API as

`void PalSetJitCpuCapabilityFlags(CORJIT_FLAGS &jitFlags);`

**Comment by RussKeldorph** [comment]
Was this fixed by #15447?  Please close if so.

**Comment by sdmaclea** [comment]
This was fixed for ARM64, ARM32 can be closed as I think it will just be implemented when needed.
